lib/spack/spack/spec.py

```diff
@@ -3438,6 +3438,12 @@ def satisfies(self, other: Union[str, "Spec"], deps: bool = True) -> bool:
             # The condition cannot be applied in any case, skip the edge
             if not self.intersects(rhs_edge.when):
                 continue
+            if (
+                rhs_edge.parent.name
+                and rhs_edge.parent.name in self
+                and not self[rhs_edge.parent.name].intersects(rhs_edge.when)
+            ):
+                continue
```

Comment:
I'll look into this. Doesn't seem entirely correct at a first glance. For instance, we already checked with `self` above. Also, as a note for the future, it would have been good to have a new unit test that was failing before the change, and succeeding after the change to avoid regressions.

Reply:
Indeed this implementation was buggy. Should be fixed in https://github.com/spack/spack/pull/50481/commits/c1acf1e2fbe3f0469024c1a6051c80f5f96c20cc that also adds a unit test